### PR TITLE
Boost: Add cache invalidation on module toggle

### LINKED ISSUE #35748: [Boost] Remember to invalidate cache when other modules are changing output
Some Boost modules can change the HTML output. Boost cache should play nicely with Boost's own modules. 

Examples:
- Critical CSS
- Minify JS/CSS
- Image CDN
- Deferred JS

### PR DESCRIPTION
**Pull request by haqadn**
Fixes #35748 

## Proposed changes:

- Invalidate cache on module toggle
- Invalidate cache on critical CSS generation
- Super Cache compatibility fix:
  - Invalidate cache only on output changing module toggle
  - Invalidate cache on C.CSS invalidation

### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [ ] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Jetpack product discussion
None

## Does this pull request change what data or activity we track or use?
No

## Testing instructions:

* Create some cache by browsing the site
* Toggle a boost module that might change the HTML output. 
* Make sure all cache got deleted. Special case critical-css is activation doesn't immidately clear cache. It waits until C.CSS generation.

### CODE REVIEW

File: projects/plugins/boost/app/contracts/Changes_Output.php
Comment (pyronaur) [inline projects/plugins/boost/app/contracts/Changes_Output.php]: Is headers a part of output? What about compression? Is that what this interface asks? Maybe we should use something like `Modifies_HTML` just so that it's specific.

File: projects/plugins/boost/app/modules/cache/Page_Cache.php
Comment (pyronaur) [inline projects/plugins/boost/app/modules/cache/Page_Cache.php]: Neat! 👏

File: projects/plugins/boost/app/modules/cache/Page_Cache.php
Comment (pyronaur) [inline projects/plugins/boost/app/modules/cache/Page_Cache.php:63]: This is fine, but I do wonder if the interface we want is actually `Affects_Speed_Score` so that this wouldn't be necessary 🤔

File: projects/plugins/boost/compatibility/wp-super-cache.php
Comment (pyronaur) [inline projects/plugins/boost/compatibility/wp-super-cache.php:61]: This makes me really want to consider `Affects_Speed_Score` as an interface name - it's very specific and unmistakeable for what it actually implies.

**Review by pyronaur — DISMISSED** [review]
This looks great, approving with suggestions - I'll leave it up to you to decide what to do with that.

File: projects/plugins/boost/app/modules/cache/Page_Cache.php
Comment (haqadn) [reply projects/plugins/boost/app/modules/cache/Page_Cache.php:63]: It does change things when we turn it of instantly. Turning on doesn't impact it immediately. If we want to differentiate it with interfaces, we may end up with: Modifies_HTML_On_Enable and Modifies_HTML_On_Disable. Which now start to look like too specific separation.

File: projects/plugins/boost/app/contracts/Changes_Output.php
Comment (haqadn) [reply projects/plugins/boost/app/contracts/Changes_Output.php]: I considered that. But I also feel like `Modifies_HTML` may imply it doesn't change inline CSS and JS. What do you think?

Comment (pyronaur) [reply projects/plugins/boost/app/contracts/Changes_Output.php]: Oh 🤔  I didn't realize it includes CSS/JS too.

Yeah. I'm not sure I can come up with a better name....

`Changes_DOM`, `Changes_Page_Output`, `Modifies_Assets_Or_HTML` 🤷  all sound pretty meh to me....

File: projects/plugins/boost/app/modules/cache/Page_Cache.php
Comment (pyronaur) [reply projects/plugins/boost/app/modules/cache/Page_Cache.php:63]: `Affects_Speed_Score` says nothing about doing it immediately 🤷

File: projects/plugins/boost/app/contracts/Changes_Output.php
Comment (haqadn) [reply projects/plugins/boost/app/contracts/Changes_Output.php]: I like `Changes_Page_Output`

**Review by pyronaur — APPROVED** [review]
We can merge as is and continue this in the next maintenance sprint

**Pull request merged**
